### DISCUSSION TITLE
HLS compatibility with Roku

### DISCUSSION THREAD

**Discussion by bchah** [OvenMediaLabs/OvenMediaEngine#1550] [Help] [answered]
I have a customer who is trying to play OME HLS links on their Roku channel through a third-party service. They can't give us any information besides "it doesn't work" 😅.

A bit of research suggested that Roku doesn't accept the Content-Type header of "application/vnd.apple.mpegurl" so I created a build of OME that lets you change the header and tried a few alternatives to no avail (PR for that coming soon 😉).

I am grasping at straws here but did notice one inconsistency... the master playlists for HLS are EXT-X-VERSION 7 but the 'level' playlist is version 10. Could this be causing a compatibility issue? Any other ideas or anecdotes from people who have tried to get OME feeding to a Roku channel?

**Comment by getroot** [comment]
I have no experience with Roku players. However, it is normal for the versions of the Master playlist and Chunk playlist to be different, and this is the correct standard.

**Comment by bchah** [comment]
When you wrote your message, I also had no experience with Roku players. Now I have one on my desk and I hate it already 😂

I can't be certain what the error is because their test tools only give us a "Video Playback failed" message. However their documentation says that they don't support fmp4 in HLS, only TS. Which then means that they don't support the HLS standard, since fmp4 is a perfectly valid segment format according to the standard.

With this understanding, is there any advice you can give as to how we can overcome this without standing up another server next to OME? Maybe the Dump feature can be configured to dump .TS instead of .mp4, the way the recording can?

**Reply by getroot** [reply, marked as answer]
ah. It only supports TS container. You can use Legacy HLS embeded in OvenMediaEngine. (Someone said it still works, but it's been a while since I've used it.)

Alternatively, I am currently developing an MPEG-2 TS packager for the SRT Publisher and the new Legacy HLS Publisher. Once this is done you will be able to use it to play on Roku.

**Reply by bchah** [reply]
@getroot you did it! I tried with OME 0.13.2 and legacy HLS works on Roku 🎉

Are you saying I can enable <HLS> publisher in applications in 0.16.x and it will create TS segments?

Exciting to hear about the new work on the SRT and legacy HLS publishers - thank you for your continued innovation!

**Reply by getroot** [reply]
Yes, I deprecated legacy HLS, but I didn't delete the code. One person told me that he was still using <Publishers><HLS> in the latest version. I think you could try it too.

The reason I deprecate it is because it was mostly temporary code. It is almost impossible to expand with drm, dvr, etc. So I'm completely remaking it.

**Reply by bchah** [reply]
Confirmed legacy HLS is working on Roku with OME 0.16.4. Brilliant. This is the power of OvenMediaEngine 👏🏼

![IMG_9341](https://github.com/AirenSoft/OvenMediaEngine/assets/33604674/ca52627a-136e-464e-93cf-83a1f46c18a3)
